File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
      //tabulation
      int n=nums.size();
      vector<int>dp(n+1,1); 
      int maxi=0;
      for(int i=0;i<n;i++){
        for(int prev=0;prev<i;prev++){
            if(nums[prev]<nums[i]) dp[i]=max(dp[prev]+1,dp[i]);
        }
        maxi=max(maxi,dp[i]);
      } 
      return maxi;
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n=len(nums)
        ans=[]
        ans.append(nums[0])
        for num in nums:
            if num>ans[-1]:
                ans.append(num)
            else:
                ans[bisect_left(ans,num)]=num #replace lowerbound element with the current element
        
        return len(ans)
        